crates/shared/src/trade_finding.rs

```diff
@@ -49,6 +120,110 @@ pub struct Trade {
     pub tx_origin: Option<H160>,
 }
 
+/// A trade with JIT orders.
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct Trade {
+    pub clearing_prices: HashMap<H160, U256>,
+    /// How many units of gas this trade will roughly cost.
+    pub gas_estimate: Option<u64>,
+    pub pre_interactions: Vec<Interaction>,
+    /// Interactions needed to produce the expected trade amount.
+    pub interactions: Vec<Interaction>,
+    /// Which solver provided this trade.
+    pub solver: H160,
+    /// If this is set the quote verification need to use this as the
+    /// `tx.origin` to make the quote pass the simulation.
+    pub tx_origin: Option<H160>,
+    pub jit_orders: Vec<dto::JitOrder>,
+}
+
+impl Trade {
+    pub fn out_amount(
+        &self,
+        buy_token: &H160,
+        sell_token: &H160,
+        in_amount: &U256,
+        order_kind: &OrderKind,
+    ) -> Result<U256> {
+        let sell_price = self
+            .clearing_prices
+            .get(sell_token)
+            .context("clearing sell price missing")?
+            .to_big_rational();
+        let buy_price = self
+            .clearing_prices
+            .get(buy_token)
+            .context("clearing buy price missing")?
+            .to_big_rational();
+        let order_amount = in_amount.to_big_rational();
+
+        let out_amount = match order_kind {
+            OrderKind::Sell => order_amount
+                .mul(&sell_price)
+                .checked_div(&buy_price)
```

Comment:
Also needs `ceil_div`

Reply:
This still does not use `ceil_div` only `checked_div`. This can lead to rounding issues which could invalidate good quotes if we configure a 0% error tolerance.